design/open/branch_checks.md

```diff
@@ -0,0 +1,171 @@
+# Branch Checks
+
+## User Story
+
+In order to achieve CI/CD for data, data engineering teams would like to validate the data itself before allowing to merge into a production branch.
+These validations might take longer than a typical HTTP request.
+
+## Goals
+
+Currently, lakeFS hooks allow running `pre-*` type hooks as blocking operations: They execute in a manner that is tied to an HTTP request.
+This puts a pretty short time window for hooks to complete in.
+
+Distributed processing frameworks such as Apache Spark typically have a long startup time and might take minutes (or even hours) to complete.
+
+The main goal of this proposal is to allow such validations to occur with a very long timeout (configurable, but could be days).
+
+## Non-Goals
+
+Out of scope for this proposal is any collaboration features that are typically associated with PRs (that provide similar functionality to the one being proposed).
+
+## High Level design
+
+This proposal adds a `Check` primitive to lakeFS branches. a Check is similar to a hook: it calls out to some external component with information about a modification or event that occurred in the system.
+
+The user flow for checks is the following:
+
+1. configure checks
+1. ensure all *mandatory* checks have passed for the given branch's HEAD
+1. merge said branch into a protected one.
+
+Merging into any protected branch requires all mandatory checks to have completed successfully for the current HEAD:
+
+```mermaid
+sequenceDiagram
+    Actor user1
+    user1->>lakeFS: commit1(branch_a)
+    user1->>lakeFS: commit2(branch_a)
+    user1->>lakeFS: commitN(branch_a)
+    user1->>lakeFS: check(branch_a) // will be dereferenced
+    loop For every configured check
+      lakeFS->>+External Executor: execute(event details, callback URL)
```

Comment:
More radical suggestion: separate concerns!  Each of these is useful on its own without necessarily all of the others.

* An _action_ starts executing when some trigger occurs: a new commit on a branch, or a new merge to the branch, or an explicit endpoint.  In future we may add new event types, just like GitHub.  This is a lot like current hooks/actions, except that it is asynchronous and there is no return value.

  Not all such actions will report a status.  Right now we call them post-* hooks IIUC, but they are equally valid as pre-* hooks.  If an action uses a reliable runner, it does not need to return a status.  For instance an "export" action just needs to trigger on commit, it's a post-commit.A "send email that a merge is starting" is a pre-merge action that does not return a status.
* An action can eventually _report status_ for a commit if it likes.  lakeFS associates this status with the commit.

  TBD: we may want to allow _anything_ to report status for a commit, regardless of whether it started as an action or not.  For example, an Airflow DAG may commit and then start validation.  It does not need lakeFS to schedule validation, indeed it may be better structured as a single DAG.  Now it can report status!
* A _protected branch_ may _require_ certain checks to pass in order to report status.